Rough draft of a style guide. Let the holy wars begin!

jeff-cohere opened the pull request:
This is all negotiable. It's just a starting point, not intended to offend anyone. :-) Please add whoever you think might care about this stuff to the list of reviewers.

Closes #118

bishtgautam commented:
It is easier to read the changes [here](https://github.com/TDycores-Project/TDycore/blob/jeff-cohere/style_guide/doc/sphinx/source/styleguide.rst).

jedbrown commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,455 @@
+For example, a constructor that converts an array of ``real_t`` to a point
+might be declared
+
+``point_t* point_from_array(real_t* array);``
```
I don't know others' opinions, but PETSc intentionally doesn't name parameters in declarations because the compiler doesn't check them and they can become stale/misleading.

I don't care in function declarations, but inside functions, I prefer
```c
real_t *a, b;
```
over
```c
real_t* a, b;
```
since the latter is often misread is suggesting that `b` would have type `real_t*`.
Style guides that recommend the `real_t* a` spacing often say only one variable can be declared per line, though that eats up precious vertical space on our luxuriously wide displays.

jeff-cohere replied:
Right on. I'll update the guide to reflect this.

ghammond86 reviewed (commented):
Looks good.  Just a few comments.

jeff-cohere commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,455 @@
+Sometimes it's convenient to provide more than one constructor, or a
+constructor that converts another datatype to a given instance of a class.
+In these cases, name each constructor so that it briefly conveys its purpose.
+For example, a constructor that converts an array of ``real_t`` to a point
```
Yep, I just noticed that this is left over stuff from the guide I copied over, which uses snake case. Sorry about that.

jeff-cohere commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,455 @@
+    // Returns the cost (in cents) of washing a load.
+    PetscInt WashingMachineCost(WashingMachine* wm);
+
+If you're familiar with contemporary object-oriented programming languages like
```
I have no strong opinion on these things either. Certainly the `GetXYZ` thing is common. In the absence of opinions, I always with "match the article of speech."

jeff-cohere commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,455 @@
+Polymorphism in C
+^^^^^^^^^^^^^^^^^
+
+**Question: do we use this idea? PETSc does, but it might not be needed in
```
Cool. If you foresee us crafting our own virtual tables, I have some text I can lift from elsewhere.

jedbrown replied:
Already in `tdycore.c`, it would be cleaner to have registration methods provide their methods instead of the switch statements.

jeff-cohere commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,455 @@
+In general, avoid comparing two floating point numbers for equality. Such
+comparisons depend on the representation of the supported precision and the
+magnitudes of the quantities under comparison. Instead...
+**Question: Any thoughts about what we encourage instead?**
```
I have used similar functions called (in this style) `RealsEqual(PetscReal, PetscReal)` and `RealsClose(PetscReal, PetscReal, PetscReal)` (with a specified tolerance).

jedbrown commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,455 @@
+
+    #include <tdycore.h>
+
+**Question**: does anyone have an opinion on the use of quotes in headers vs the
```
They have different semantics. If you want to find a header that lives in the same directory and has not been added to the header search path (by `-I.` or similar), then use quotes. Otherwise use angle brackets to skip that search of the current directory.

bishtgautam commented on doc/sphinx/source/styleguide.rst:
```diff
@@ -1,4 +1,476 @@
+
+    PetscErrorCode TDyCreateWashingMachine(PetscInt numCents, TDyWashingMachine *wm);
+
+**Question: I've seen TDycore code with Create up front, in contrast to
```
We should change TDycore to follow PETSc convention

bishtgautam commented:
@jedbrown Is there a libCEED style guide?

jedbrown commented:
libCEED uses `astyle` and `clang-tidy` (also run via CI) for style/coding standards. We don't have a formal document on the object model, but the number of types is relatively small.

jeff-cohere commented:
Thanks for the feedback, everyone. I've added a section on polymorphism, updated the section on constructors, and floating point number comparisons (functions described to be implemented). If no one objects, I'll merge this PR today. We can always revisit anything in this guide.

nocollier commented:
I am very late to this game, but I had a few rough comments:

* I do not have a personal sense of what style should be adhered to. Nor
have I ever read a style guide. I mostly just try to stick to the style I
see around me.
* I like this guide because it explains why some things are the way they
are. I learned a lot about coding just from repeating what I saw others do,
but the underlying reasons often make sense. They just aren't clear from
the outset. A guide like this I think has the possibility to bring others
along more quickly. Keeping it relatively brief is key.
* I am clearly in the PETSc camp. I admire their work and feel they have
(mostly?) good reasons for what they do. I am happy to blindly adopt
whatever they do.


On Fri, Nov 13, 2020 at 5:29 PM Jeffrey N. Johnson <notifications@github.com>
wrote:

> Merged #119 <https://github.com/TDycores-Project/TDycore/pull/119> into
> master.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/TDycores-Project/TDycore/pull/119#event-3995608435>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKFCB4ORARQ4SPUYYMBIZLSPWXLFANCNFSM4TRFCLQA>
> .
>

jeff-cohere commented:
Thanks, @nocollier . Are there any changes you would suggest to what's written so far, or anything you feel could be better explained?
